cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
     }
 
     Set<String> remainingSources = Sets.intersection(remainingNodes, possibleNewSources);
+    Set<String> processedNodes = new HashSet<>();
+    // Since there can be remaining sources from subdags which don't overlap, they should be split as seperate subdags.
+    // For that, we first create all subdags from remaining sources and keep track of accessible nodes for those
+    // subdags. Then we merge overlapping subdags
     if (!remainingSources.isEmpty()) {
-      dags.add(subsetFrom(remainingSources, possibleNewSinks));
+      // source -> [ nodes accessible by the source ]
+      Map<String, Set<String>> nodesAccessibleBySources = new HashMap<>();
+
+      for (String remainingSource : remainingSources) {
+        Dag remainingNodesDag = subsetFrom(ImmutableSet.of(remainingSource), possibleNewSinks);
+        nodesAccessibleBySources.put(remainingSource, accessibleFrom(remainingSource, remainingNodesDag.getSinks()));
+      }
+
+      for (String remainingSource : remainingSources) {
+        if (!processedNodes.contains(remainingSource)) {
+          Set<String> remainingAccessibleNodes = nodesAccessibleBySources.get(remainingSource);
+          // Island of all the remaining nodes
+          Set<String> islandNodes = new HashSet<>();
+          islandNodes.addAll(remainingAccessibleNodes);
+          for (String otherSource : remainingSources) {
+            if (!remainingSource.equalsIgnoreCase(otherSource)) {
+              Set<String> otherAccessibleNodes = nodesAccessibleBySources.get(otherSource);
+              // If there is an overlap, add those nodes to the island.
+              if (!Sets.intersection(remainingAccessibleNodes, otherAccessibleNodes).isEmpty()) {
+                islandNodes.addAll(otherAccessibleNodes);
+              }
+            }
+          }
+          dags.add(createDag(islandNodes));
+          // keep track of processed nodes
+          processedNodes.addAll(islandNodes);
+        }
+      }
     }
     return dags;
   }
 
+  /**
+   * Creates dag from provided list of island nodes
+   * @param islandNodes list of island nodes
+   * @return Dag with connections among island nodes
+   */
+  private Dag createDag(Set<String> islandNodes) {
```

Comment:
It would be better to move this into the Dag class as `createSubDag(Set<String> nodes)`, since it is creating another Dag from a subset of the nodes in the current Dag. 'island' doesn't really describe anything.